### PR TITLE
STAC-12961: WIP: Add documentation for health sync agent 

### PR DESCRIPTION
**Pull request by craffit**

### CODE REVIEW

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 1]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+
+```
+{% endtab %}
+{% endtabs %}
+
+The `HealthStream` class has the following options:
+
+* **urn** - HealthStreamUrn. The stream urn under which the health infromation will be grouped.
+* **sub_stream** - string. Optional. Allows for separating disjoint data sources within a single health synchronization stream. (E.g. the data for the streams is reprted separately form different hosts)
+* **repeat_interval_seconds** - integer. Optional. The interval with which data will be repeated, defaults to `min_collection_interval`. This allows stackstate to detect when data is 'late'.
```
Comment (gggina) [inline develop/developer-guides/agent_check/agent-check-api.md:270]: StackState

why are there quotes around `late` ? is this an actual name that is used somewhere or should we better say "when data arrives later than expected".

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 2]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+* **expiry_seconds** - integer. Optional. The time after which all data form the (sub)stream should be removed. Defaults to 4 times the repeat interval. Expiry can be disabled by passing '0', if a sub_stream is not specified.
```
Comment (gggina) [inline develop/developer-guides/agent_check/agent-check-api.md:271]: * **expiry_seconds** - integer. Optional. The time after which all data from the stream or substream should be removed. Set to '0' to disable expiry. Default 4*`repeat_interval_seconds`. 

>> I didn't understand the last section "if a sub_stream is not specified."

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 3]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+
+
+More information on urns, health synchronization streams, snapshots and how to debug can be found in the platform docs for [health Synchronization](/configure/health/health-synchronization.md).
```
Comment (gggina) [inline develop/developer-guides/agent_check/agent-check-api.md:274]: For more information on urns, health synchronization streams, snapshots and how to debug, see [health Synchronization](/configure/health/health-synchronization.md).

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 4]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+
+```
+{% endtab %}
+{% endtabs %}
+
+The method requires the following details:
+
+* **check_state_id** - string. Uniquely identifies the check state within the (sub)stream.
+* **name** - string. Display name for the health check state.
+* **health_value** - Health. The stackstate health value, can be CLEAR, DEVIATING or CRITICAL.
```
Comment (gggina) [inline develop/developer-guides/agent_check/agent-check-api.md:304]: StackState

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 5]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+* **topology_element_identifier** - string. The component or relation identifier the check state should bind to. The check state will associated with all components/realtions which have the specified identifier.
```
Comment (gggina) [inline develop/developer-guides/agent_check/agent-check-api.md:305]: The component or relation identifier that the check state should bind to. The check state will associated with all components/relations that have the specified identifier.

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 6]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+* **message** - string. Optional. Extended message do display with the health state. Supports Markdown.
```
Comment (gggina) [inline develop/developer-guides/agent_check/agent-check-api.md:306]: * **message** - string. Optional. Extended message to display together with the health state. Supports Markdown.

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 7]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+
+See the example of creating a component in StackState in the [StackState Static Health \(github.com\)](https://github.com/StackVista/stackstate-agent-integrations/blob/master/static_health/stackstate_checks/static_health/static_health.py).
```
Comment (gggina) [inline develop/developer-guides/agent_check/agent-check-api.md:308]: >> again avoiding gerund

For an example of how to create a component, see the [StackState Agent Static Health check \(github.com\)](https://github.com/StackVista/stackstate-agent-integrations/blob/master/static_health/stackstate_checks/static_health/static_health.py).

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md [hunk 8]
```diff
@@ -239,6 +239,61 @@ self.event({
...
+        if 'url' not in instance:
+            raise ConfigurationError('Missing url in topology instance configuration.')
+        instance_url = instance['url']
+        return HealthStream(HealthStreamUrn("example", instance_url))
+
+...
+
+```
+
+This function specifies what health synchronization stream this agent check will produce to. Having this function defined will enable the `self.health` api.
```
Comment (gggina) [inline develop/developer-guides/agent_check/how_to_develop_agent_checks.md:267]: Agent check
API

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md [hunk 9]
```diff
@@ -239,6 +239,61 @@ self.event({
...
+
+### Health Synchronization Snapshots
+
+Like with topology, health data is presented to StackState using snapshots, to allow for removal of health information (check states) when they do not exist anymore in the source monitoring system. Creating snapshots is facilitated by two functions:
```
Comment (gggina) [inline develop/developer-guides/agent_check/how_to_develop_agent_checks.md:271]: >> Keep sentences short

Like with topology, health data is presented to StackState using snapshots. This allows for removal of health information (check states) when they no longer exist in the source monitoring system. Snapshots are created by two functions:

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md [hunk 10]
```diff
@@ -239,6 +239,61 @@ self.event({
...
+
+* `self.health.start_snapshot()` - used to start a health snapshot.
+* `self.health.stop_snapshot()` - used to stop the snapshot, signalling that all submitted data is complete. This should be done at the end of the check, after all data has been submitted. If exceptions occur in the check, or for some other reason not all data can be produced, this function should not be called.
```
Comment (gggina) [inline develop/developer-guides/agent_check/how_to_develop_agent_checks.md:274]: used to stop the snapshot, signaling that all submitted data is complete. This should be done at the end of the check, after all data has been submitted. If exceptions occur in the check, or for some other reason not all data can be produced, this function should not be called.

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md [hunk 11]
```diff
@@ -239,6 +239,61 @@ self.event({
...
+
+### Send Check State
+
+Check states can be send through the health synchronization api, using the `self.health.check_state()` functions in the `AgentCheck` interface. The example below shows how to submit the data:
```
Comment (gggina) [inline develop/developer-guides/agent_check/how_to_develop_agent_checks.md:278]: Check states can be sent through the health synchronization API using the `self.health.check_state()` functions in the `AgentCheck` interface. The example below shows how to submit the data:

In the example below, a check state is created in StackState with the health value `CRITICAL`. 

* The check is attached to the component or relation matching the `topology_element_identifier`. 
* The `check_state_id` is used to distinguish check states within the current health stream.

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md [hunk 12]
```diff
@@ -239,6 +239,61 @@ self.event({
...
+  check_state_id="check_state_from_example_1",
+  name="Example check state",
+  health_value=Health.CRITICAL,
+  topology_element_identifier="urn:component/the_component_to_attach_to",
+  message="Optional clarifying message"
+)
+
+```
+
+This creates a check state in the StackState product with the health value `CRITICAL`, attached to the component or relation matching the `topology_element_identifier`. The `check_state_id` is used to distinguish check states within the current health stream. If after following the previous steps you health data does not show in StackState, there is a [troubleshooting guide](/configure/health/debug-health-sync.md) for the health synchronization.
```
Comment (gggina) [inline develop/developer-guides/agent_check/how_to_develop_agent_checks.md:295]: If after following the previous steps you health data does not show in StackState, there is a [troubleshooting guide](/configure/health/debug-health-sync.md) for the health synchronization.

>> other info added above example

File: stackpacks/integrations/static_health.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+       health_file: /path/to/health.csv
+       delimiter: ';'
+     ```
+
+2. [Restart the StackState Agent\(s\)](agent.md#start-stop-restart-the-stackstate-agent) to apply the configuration changes.
+3. Once the Agent has restarted, wait for the Agent to collect data from the specified [health CSV file](#csv-file-format) and send it to StackState.
+
+### Validate
+
+Execute the info command and verify that the integration check has passed. The output of the command should contain a section similar to the following:
```
[the remark this answers is not in the document]
Comment (craffit) [reply stackpacks/integrations/static_health.md:42]: that command seems to not exist anymore, so i'll also drop this section

File: stackpacks/integrations/static_health.md [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+The delimiter used in the CSV file can be specified when you [configure the Static Health check](#configure) on the StackState Agent.
+
+
+## Release notes
+
+## See also
+
+* [StackState Agent V2](agent.md) 
+* [Health state in StackState](health-state-and-event-notifications/health-state-in-stackstate.md)
```
[the remark this answers is not in the document]
Comment (gggina) [reply stackpacks/integrations/static_health.md:87]: >> Ran a link check on the branch and this link is broken. It should be: 

[Health state in StackState](/use/health-state-and-event-notifications/health-state-in-stackstate.md)

File: stackpacks/integrations/static_health.md [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+### Delimiter
+
+The delimiter used in the CSV file can be specified when you [configure the Static Health check](#configure) on the StackState Agent.
+
+
+## Release notes
+
+## See also
+
+* [StackState Agent V2](agent.md)
```
[the remark this answers is not in the document]
Comment (craffit) [reply stackpacks/integrations/static_health.md:86]: I added links to health sync overview and health sync debug

File: stackpacks/integrations/static_health.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (craffit) [reply stackpacks/integrations/static_health.md:87]: fixed

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md (hunk 8 above)
Comment (craffit) [reply develop/developer-guides/agent_check/how_to_develop_agent_checks.md:267]: updated

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md (hunk 9 above)
Comment (craffit) [reply develop/developer-guides/agent_check/how_to_develop_agent_checks.md:271]: copy-pasted

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md (hunk 10 above)
Comment (craffit) [reply develop/developer-guides/agent_check/how_to_develop_agent_checks.md:274]: copy-pasted

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md (hunk 11 above)
Comment (craffit) [reply develop/developer-guides/agent_check/how_to_develop_agent_checks.md:278]: copy-pasted (and deduplicated the `example below` part

File: develop/developer-guides/agent_check/how_to_develop_agent_checks.md (hunk 12 above)
Comment (craffit) [reply develop/developer-guides/agent_check/how_to_develop_agent_checks.md:295]: copy-pasted, introducing the link

File: configure/health/health-synchronization.md [hunk 16]
```diff
@@ -6,7 +6,7 @@ Health synchronization allows you to add existing health checks from external mo
 
 
 ## Set up health synchronization
-The StackState receiver API will automatically receive and process all incoming health data. No additional configuration is required in StackState to enable this, however, the health data received should match the expected JSON format. For details, see [send health data over HTTP](/configure/health/send-health-data.md).
+The StackState receiver API will automatically receive and process all incoming health data. No additional configuration is required in StackState to enable this, however, the health data received should match the expected JSON format. For details on how to ingest through the api, see [send health data over HTTP](/configure/health/send-health-data.md), for ingesting using the agent, see [how to develop agent checks](/develop/developer-guides/agent_check/how_to_develop_agent_checks.md).
```
[the remark this answers is not in the document]
Comment (craffit) [reply configure/health/health-synchronization.md:9]: updated

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 17]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
 
 Check the usage in the following [example](https://github.com/StackVista/stackstate-agent-integrations/blob/master/mysql/stackstate_checks/mysql/mysql.py#L434).
 
+### Health
+
+Health information can be sent to StackState with the following methods:
+
+* `self.health.check_state` - send a check state as part of a snapshot.
+* `self.health.start_snapshot()` - used to start a health snapshot. Stackstate only processes health information if it is sent as part of a snapshot.
```
[the remark this answers is not in the document]
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:231]: updated

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 18]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+* `self.health.stop_snapshot()` - used to stop the snapshot, signalling that all submitted data is complete. This should be done at the end of the check, after all data has been submitted. If exceptions occur in the check, or for some other reason not all data can be produced, this function should not be called.
```
[the remark this answers is not in the document]
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:232]: updated

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 19]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+
+#### Setting up a health stream
```
[the remark this answers is not in the document]
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:234]: updated

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 20]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+
+To make the `self.health` api avilable, we need to define a that we associate the health information with. This can be done by overriding the `get_health_stream` function, as in the following example:
```
[the remark this answers is not in the document]
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:236]: I made it: To make the `self.health` API available, override the `get_health_stream` function to define a URN identifier for the health synchronization stream.

File: develop/developer-guides/agent_check/agent-check-api.md [hunk 21]
```diff
@@ -223,6 +223,89 @@ The method can accept the following arguments:
...
+...
+
+```
+{% endtab %}
+{% endtabs %}
+
+The `HealthStream` class has the following options:
+
+* **urn** - HealthStreamUrn. The stream urn under which the health infromation will be grouped.
+* **sub_stream** - string. Optional. Allows for separating disjoint data sources within a single health synchronization stream. (E.g. the data for the streams is reprted separately form different hosts)
```
[the remark this answers is not in the document]
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:269]: updated, good tips!

File: develop/developer-guides/agent_check/agent-check-api.md (hunk 1 above)
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:270]: i made it: 'when data arrives later than expected'

File: develop/developer-guides/agent_check/agent-check-api.md (hunk 2 above)
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:271]: updated and tried to include the sub_stream part

File: develop/developer-guides/agent_check/agent-check-api.md (hunk 3 above)
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:274]: updated

File: develop/developer-guides/agent_check/agent-check-api.md (hunk 4 above)
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:304]: updated

File: develop/developer-guides/agent_check/agent-check-api.md (hunk 5 above)
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:305]: updated

File: develop/developer-guides/agent_check/agent-check-api.md (hunk 6 above)
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:306]: updated

File: develop/developer-guides/agent_check/agent-check-api.md (hunk 7 above)
Comment (craffit) [reply develop/developer-guides/agent_check/agent-check-api.md:308]: updated

**Review by gggina — APPROVED** [review]

**Pull request merged**